Fixes Series comparison to NumPy scalars

jrbourbeau opened the pull request:
Fixes #3392. This PR fixes Series comparison to NumPy scalars, which currently fail.

- [x] Tests added / passed
- [x] Passes `flake8 dask`
- [x] Fully documented, including `docs/source/changelog.rst` for all changes
      and one of the `docs/source/*-api.rst` files for new API

mrocklin commented:
Thank you for taking this on @jrbourbeau !

@philippjfr does this resolve your original issue?

jrbourbeau commented:
The original regression test I added was using the `dask.array.greater_equal` ufunc, which actually doesn't raise any errors. So I replaced the added test with the original code snippet given in issue #3392.

mrocklin commented:
It looks like there are some genuine testing errors in travis-ci.  I suspect that this fix works for some versions, but not others.

jrbourbeau commented:
It looks like the NumPy 1.11.0 tests on Travis are failing because I've only modified `__array_ufunc__` (which is new in NumPy version 1.13) and not `__array_wrap__` (which is what's being used in the 1.11.0 tests).

jrbourbeau commented:
Looks like there was an issue during the installation process on some of the Travis tests (see https://travis-ci.org/dask/dask/jobs/371219552#L827-L832). Could someone please restart the failed builds?

mrocklin commented:
Done

On Wed, Apr 25, 2018 at 4:36 PM, James Bourbeau <notifications@github.com>
wrote:

> Looks like there was an issue during the installation process on some of
> the Travis tests (see https://travis-ci.org/dask/
> dask/jobs/371219552#L827-L832). Could someone please restart the failed
> builds?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/3436#issuecomment-384425691>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AASszC1peBFOtr0vwRmaY6xgsQafCqjZks5tsN5UgaJpZM4ThuME>
> .
>

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -344,8 +344,11 @@ def __array_wrap__(self, array, context=None):
         for x in inputs + out:
-            if not isinstance(x, (Number, Scalar, _Frame, Array,
-                                  pd.DataFrame, pd.Series, pd.Index)):
+            # Want to check for 0-dimensional arrays
```
Nitpick: it'd be nice if the comment explained why we're doing this, instead of what we're doing.

jrbourbeau replied:
That's a fair point. I can update it to something like: 
```python
# ufuncs work with 0-dimensional NumPy ndarrays
# so we don't want to raise NotImplemented
```

TomAugspurger commented on dask/dataframe/core.py:
```diff
@@ -1751,7 +1754,10 @@ class Series(_Frame):
     def __array_wrap__(self, array, context=None):
         if isinstance(context, tuple) and len(context) > 0:
-            index = context[1][0].index
+            if isinstance(context[1][0], np.ndarray) and context[1][0].shape == ():
```
This is getting *close* to enough code where we might want to refactor it to a base method. In pandas this would be a `_shallow_copy`. You pass it the data, and the result is constructed with all the metadata and indexes.

For now though, it's probably fine.

jrbourbeau commented:
> Does the same wrapping need to be applied to `dd.Index`?

Just to clarify, do you mean that `dd.Index.__array_wrap__` should also have a check for 0-dimensional ndarrays? Currently, if a 0-dimensional ndarray is given to `dd.Index.__array_wrap__` it raises 

```python
TypeError: len() of unsized object
```

martindurant commented:
How about `(np.array(1) + df.index).compute()`

TomAugspurger commented:
Ah that return TypeError, from a NotImplemented.

On Thu, Apr 26, 2018 at 1:54 PM, Martin Durant <notifications@github.com>
wrote:

> How about (np.array(1) + df.index).compute()
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/3436#issuecomment-384752903>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABQHIvatWdzmh5BdJaXy9N2mSadSlMMQks5tshf4gaJpZM4ThuME>
> .
>

jrbourbeau commented:
Hmmmm I'm able to reproduce Tom's example. I.e.

```python
import numpy as np
import dask.dataframe as dd
import pandas as pd

df = dd.from_pandas(pd.Series([1, 2, 3], index=[0, 2, 3]), 2)

print((df.index + np.array(1)).compute())
```

Gives `Int64Index([1, 3, 4], dtype='int64')`. 

But reversing the addition order changes behavior depending on which numpy version I'm using. 

- With numpy 1.11.0, `np.array(1) + df.index` returns a pandas Index `Int64Index([1, 3, 4], dtype='int64')`

-  With numpy 1.14.2, `np.array(1) + df.index` returns a dask Index. After `compute()` is called it gives `Int64Index([1, 3, 4], dtype='int64')`

I take it this difference has to do with the fact that `__array_ufunc__` is using `elemwise` while `__array_wrap__` simply returns a pandas Index, i.e. `pd.Index(array, name=self.name)`.

mrocklin commented:
Checking in here.  What is the status of this?

TomAugspurger commented:
Haven't looked closely, but I think everything I looked at last time was fine. I'd be happy with merging this as is, and opening a separate issue to track the (presumably less common) ops on the Index class, if there's even an issue there.

jrbourbeau commented:
Just updated comment as per Tom's suggestion

mrocklin commented:
Great.  This is in. Thanks @jrbourbeau !
